docs/contributing/contributing.md

```diff
@@ -72,11 +65,52 @@ To create binaries:
 
 GHC 8.6.5 is not supported here because `nixpkgs-unstable` no longer maintains the corresponding packages set.
 
-## Introduction tutorial
+## Testing
 
-See the [tutorial](./plugin-tutorial.md) on writing a plugin in HLS.
+The tests make use of the [Tasty](https://github.com/feuerbach/tasty) test framework.
+
+There are two test suites in the main haskell-language-server package, functional tests, and wrapper tests.
+Other project packages, like the core library or plugins, can have their own test suite.
+
+### Testing with Cabal
+
+Running all the tests
+
+```bash
+$ cabal test
+```
+
+Running just the functional tests
+
+```bash
+$ cabal test func-test
+```
+
+Running just the wrapper tests
+
+```bash
+$ cabal test wrapper-test
+```
+
+Running a subset of tests
+
+Tasty supports providing
+[Patterns](https://github.com/feuerbach/tasty#patterns) as command
+line arguments, to select the specific tests to run.
+
+```bash
+$ cabal test func-test --test-option "-p hlint"
+```
 
-## Test your hacked HLS in your editor
+The above recompiles everything every time you use a different test option though.
+
+An alternative, which only recompiles when tests (or dependencies) change:
+
+```bash
+$ cabal run haskell-language-server:func-test -- -p "hlint enables"
```

Comment:
No, that won't help. `cabal run` does not change the environment, as far as I know. 
You can do `cabal exec cabal run h-l-s:func-test -- -- -p ...` and it will work sometimes

Reply:
> 
> 
> No, that won't help. `cabal run` does not change the environment, as far as I know.

It seems so: https://github.com/haskell/cabal/issues/5411#issuecomment-402548725. It works for `cabal test`but no for `cabal run` iiuic that comment

> You can do `cabal exec cabal run h-l-s:func-test -- -- -p ...` and it will work sometimes

It works enough times to update the instructions? are you are referring to the fact `cabal exec` will not rebuild the executable and you could end using a older version?

Reply:
Moreover with cabal-install-3.8, test arguments will not trigger a rebuild

Reply:
I believe `cabal exec` does trigger a rebuild, but it fails with weird errors in some cases.